pykafka/simpleconsumer.py

```diff
@@ -646,20 +655,18 @@ def _handle_success(parts):
                         min_bytes=self._fetch_min_bytes
                     )
                 except (IOError, SocketDisconnectedError):
+                    if self._running:
+                        unlock_partitions(iterkeys(partition_reqs))
+                        self._update()
                     # If the broker dies while we're supposed to stop,
                     # it's fine, and probably an integration test.
-                    if self._running:
-                        raise
                     return
-
                 parts_by_error = build_parts_by_error(response, self._partitions_by_id)
                 handle_partition_responses(
                     self._default_error_handlers,
                     parts_by_error=parts_by_error,
                     success_handler=_handle_success)
-                # unlock the rest of the partitions
-                for owned_partition in iterkeys(partition_reqs):
-                    owned_partition.fetch_lock.release()
+                unlock_partitions(iterkeys(partition_reqs))
```

Comment:
The same as in the other comment will apply here. I find it really difficult to track what's locked and what's not, so I'm wondering if maybe - even if that doesn't look very elegant - it would help to just keep an explicit `list()` of fetch-locks held.